x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/ingest/SetSecurityUserProcessor.java

```diff
@@ -85,6 +86,39 @@ public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
                         userObject.put("metadata", user.metadata());
                     }
                     break;
+                case API_KEY:
+                    final HashMap<String, Object> apiKey = new HashMap<>();
+                    Object apiKeyName = authentication.getMetadata().get(ApiKeyService.API_KEY_NAME_KEY);
+                    if (apiKeyName != null) {
+                        apiKey.put("name", apiKeyName);
+                    }
+                    Object apiKeyId = authentication.getMetadata().get(ApiKeyService.API_KEY_ID_KEY);
+                    if (apiKeyId != null) {
+                        apiKey.put("id", apiKeyId);
+                    }
+                    if (false == apiKey.isEmpty()) {
+                        userObject.put("api_key", apiKey);
+                    }
+                    break;
+                case REALM:
+                    final Object realmName = ApiKeyService.getCreatorRealmName(authentication);
+                    final Object realmType = ApiKeyService.getCreatorRealmType(authentication);
```

Comment:
I think there's a mismatch here.

The API Key service populates the API key creator realm based on the _looked-up-by_.
But the `getCreatorRealm*` methods return the authenticating realm, not the looked up realm.

So, if I do `curl -u elastic -H "es-security-runas-user: foo" PUT /index/_doc/1` the pipeline will record the realm as `reserved`

But if I do `curl -u elastic -H "es-security-runas-user: foo" PUT /_security/api_key ...` and then `curl -H "Authorization: ApiKey .... PUT /index/_doc/1` it will have a realm of `native`.

It's not a big deal, but it's inconsistent. I think we want to always use the _looked-up-by_ in the ingest processor, maybe with a flag of `lookup: true` or something like that if the authenticated realm was different.

Reply:
You are right. I think the first case should report realm as `native` as well. This should be easily achieved once we augument the `Authentication` class with a new method to resolve the actual realm.

As for the `lookup: true` flag, do you mean add it as part of the SetSecurityUserProcessor configuration?

Reply:
Updated for both [name](https://github.com/elastic/elasticsearch/pull/51305/files#diff-cc4f9d4c564aae71214ac709ed789b4dR887) and [type](https://github.com/elastic/elasticsearch/pull/51305/files#diff-cc4f9d4c564aae71214ac709ed789b4dR902).